File: leetcode-go/Q051-Q100/Q099_Recover_Binary_Search_Tree/Q099_Recover_Binary_Search_Tree.go
package Q099_Recover_Binary_Search_Tree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func recoverTree(root *TreeNode) {
	var lnode, rnode, pnode *TreeNode

	var traverse func(root *TreeNode)
	traverse = func(root *TreeNode) {
		if root == nil {
			return
		}
		traverse(root.Left)
		if pnode != nil && pnode.Val >= root.Val {
			if lnode == nil {
				lnode = pnode
			}
			rnode = root
		}
		pnode = root
		traverse(root.Right)
	}

	traverse(root)
	lnode.Val, rnode.Val = rnode.Val, lnode.Val
	return
}

File: src/Y2018/facebook/Q099_Recover_Binary_Search_Tree.java
package Y2018.facebook;

public class Q099_Recover_Binary_Search_Tree {

    TreeNode first = null;
    TreeNode second = null;
    TreeNode prev = null;

    public void recoverTree(TreeNode root) {
        dfs(root);
        int tmp = first.val;
        first.val = second.val;
        second.val = tmp;
    }

    private void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        dfs(root.left);

        if (first == null && prev != null && prev.val > root.val) {
            first = prev;
        }
        if (first != null && prev != null && prev.val > root.val) {
            second = root;
        }
        prev = root;

        dfs(root.right);
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

}
